src/logger.ts

```diff
@@ -21,6 +21,126 @@ import type { LoggingConfig } from "./types.ts";
 import { parse } from "@std/jsonc";
 import { dirname, join, resolve } from "@std/path";
 
+// ===== Hex Formatter Utilities =====
+
+/**
+ * Convert byte array to compact hex string (no spaces)
+ * @param data Uint8Array, ArrayBuffer, or number array
+ * @param maxBytes Optional maximum bytes to output (truncates if exceeded, no limit by default)
+ * @returns Hex string like "deadbeef"
+ */
+function toHexCompact(data: Uint8Array | ArrayBuffer | number[], maxBytes?: number): string {
+  if (!data) return String(data); // Handle null/undefined
+
+  try {
+    const bytes = data instanceof ArrayBuffer ? new Uint8Array(data) : Array.isArray(data) ? new Uint8Array(data) : data;
+
+    // Truncate if maxBytes specified and exceeded
+    if (maxBytes !== undefined && bytes.length > maxBytes) {
+      const truncated = Array.from(bytes.slice(0, maxBytes))
+        .map((b) => b.toString(16).padStart(2, "0"))
+        .join("");
+      return `${truncated}... [truncated, ${bytes.length - maxBytes} more bytes]`;
+    }
+
+    return Array.from(bytes)
+      .map((b) => b.toString(16).padStart(2, "0"))
+      .join("");
+  }
+  catch {
+    // Fallback for wrong types
+    return String(data);
+  }
+}
+
+/**
+ * Convert byte array to spaced hex string (with spaces for readability)
+ * @param data Uint8Array, ArrayBuffer, or number array
+ * @param maxBytes Optional maximum bytes to output (truncates if exceeded, no limit by default)
+ * @returns Hex string like "de ad be ef"
+ */
+function toHexSpaced(data: Uint8Array | ArrayBuffer | number[], maxBytes?: number): string {
+  if (!data) return String(data);
+
+  try {
+    const bytes = data instanceof ArrayBuffer ? new Uint8Array(data) : Array.isArray(data) ? new Uint8Array(data) : data;
+
+    // Truncate if maxBytes specified and exceeded
+    if (maxBytes !== undefined && bytes.length > maxBytes) {
+      const truncated = Array.from(bytes.slice(0, maxBytes))
+        .map((b) => b.toString(16).padStart(2, "0"))
+        .join(" ");
+      return `${truncated} ... [truncated, ${bytes.length - maxBytes} more bytes]`;
+    }
+
+    return Array.from(bytes)
+      .map((b) => b.toString(16).padStart(2, "0"))
+      .join(" ");
```

Comment:
So... the only difference between `toHexSpaced` and `toHexCompact` is this join(" ")?

DRY?